### PR TITLE
refactor(encoding): #111 Phase 1e — migrate methods off HcMatchGenerator (Stages A–D)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 1e of the #111 encoder architecture rewrite. Migrates the bulk of
`impl HcMatchGenerator` off the monolith and onto the per-backend impls
that Phase 1d (#118) carved out. All Stages A / B / C / D land here.

22 batches × ~95 methods + 5 exported macros + 3 const moves +
1 enum introduction. Each batch keeps:

  - `cargo build` (default + `--no-default-features`)
  - `cargo clippy --all-targets -- -D warnings`
  - `cargo nextest run -p structured-zstd --lib` → 381 / 381 passed
  - `level22_sequences_match_donor_on_corpus_proxy` ratio gate

green at every commit.

## Where everything ended up

**`impl MatchTable`** now owns every hash / chain / BT-tree byte:
- Hash family (`ensure_tables`, `hash_position*`, `hash_value_with_mls`,
  `hash3_position`, `read_le_u32*`), insert family
  (`insert_position`, `insert_positions[_with_step]`, `*_no_rebase`,
  `update_hash3_until`), rebase guard chain
  (`maybe_rebase_positions`, `rebase_positions_cold`,
  `replay_history_for_rebase_hc`, `replay_history_for_rebase_bt`,
  `begin_rebase`, `needs_rebase`, `can_skip_rebase_check_at`).
- BT walker — `bt_insert_step_no_rebase` ×5 SIMD + dispatcher,
  `bt_insert_and_collect_matches` ×5 SIMD + dispatcher,
  `bt_update_tree_until` ×5 SIMD + dispatcher,
  `bt_insert_sparse_incompressible_block`.
- HC3 short-match probe — `hash3_candidate` ×5 SIMD + dispatcher.
- Slice-boundary glue — `backfill_boundary_positions`,
  `apply_limited_update_after_long_match`, `skip_matching`,
  `emit_optimal_plan`.

**`impl HcMatcher`** owns HC chain-walk methods (`find_best_match`,
`pick_lazy_match`, `chain_candidates`, `repcode_candidate`,
`hash_chain_candidate`, `extend_backwards`, `match_gain`,
`better_candidate`, `for_each_repcode_with_reps_*` ×5 SIMD,
`sufficient_match_len_for_pass`).

**`impl BtMatcher`** owns BT-side optimal-parser helpers
(`push_candidate_ladder`, LDM helpers, `finish_optimal_plan`,
`prepare_ldm_candidates`, optimal-parser pure helpers
`encode_offset_*`, `update_plan_stats_segment`, `reset_opt_node*`,
`add_price*`, `cached_*_price` × 4).

**`HcMatchGenerator`** is now thin orchestration only — configure /
reset / seed_dictionary_entropy / skip_matching wrapper /
start_matching (lazy + optimal) / build_optimal_plan family /
run_btultra2_seed_pass. Stage D's
`enum HcBackend { Hc, Bt(Box<BtMatcher>) }` means HC modes
(levels 1–15) no longer allocate the ~4 KiB BT optimal-parser
scratch payload.

## Performance check

level22 compress, pure_rust median (criterion `--quick`), before vs
after Stage D's enum + Box indirection (batch 21 → batch 22):

| Scenario              | Pre-enum | Post-enum | Δ        |
| ---                   | ---      | ---       | ---      |
| decodecorpus-z000033  | 478 ms   | 471 ms    | -1.3 %   |
| high-entropy-1m       | 3.62 ms  | 3.56 ms   | -1.7 %   |
| low-entropy-1m        | 1.37 ms  | 1.38 ms   | +0.5 %   |
| large-log-stream      | 136 ms   | 132 ms    | -3.2 %   |

All deltas within criterion quick-mode noise band (±5 %), no
regression from the `Box<BtMatcher>` indirection on the BT hot path.

## Codecov note

`codecov/patch` reports ~72 % patch coverage versus the project's ~90 %
target. Treat the gate as informational on this PR — it is intrinsic to
the move-only nature of the change, not a coverage regression:

- This PR moves ~95 methods between files (match_generator.rs → match_table/storage.rs / bt/mod.rs / hc/mod.rs). No behaviour changes; 381 / 381 tests pass including the level22 ratio gate.
- The bulk of "uncovered" lines are SIMD walker / probe variants (`*_neon`, `*_sse42`, `*_avx2_bmi2`, `*_scalar`) that are runtime-selected per CPU. CI exercises one kernel at a time, so the same lines were uncovered before this PR — they only show up in `codecov/patch` now because they got re-credited as "new" when moved between files.
- Coverage of the actual orchestration path (`HcMatchGenerator`, encode-side entry points) is unchanged.

## Commits

23 commits (batch 1 → batch 22 + 1 prep). Each batch is a
self-contained mechanical move + caller rewrite; commit messages list
the methods moved, the migration pattern used, and the verification
matrix output.

Part of #111 — completes Phase 1e (Stages A–D). The umbrella issue stays open for follow-on phases.

## Summary by CodeRabbit

* **New Features**
  * Long-distance match candidate generation and expanded optimal-parser/encoder support.

* **Refactor**
  * Major internal overhaul of match-finding, table/storage and optimal-planning plumbing with multi-backend dispatch.

* **Performance**
  * Faster short- and long-match probing via architecture-specific kernels and improved hashing/indexing.

* **Stability**
  * Improved history/window/rebase handling and edge-case protections.

* **Tests**
  * New unit tests for LDM, match-finder bounds, window/seq-store behaviors.

* **Chores**
  * CI: test workflow adjusted to avoid early cancellation and use stable toolchain invocation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/119)

### CODE REVIEW

File: zstd/src/encoding/match_table/storage.rs [hunk 1]
```diff
@@ -133,4 +498,881 @@ impl MatchTable {
...
+        }
+
+        let mut literals_start = 0usize;
+        for item in plan {
+            let lit_len = item.lit_len as usize;
+            let match_len = item.match_len as usize;
+            let start = literals_start.saturating_add(lit_len);
+            if start < literals_start || start + match_len > current_len {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_table/storage.rs]: @coderabbitai Fixed — dropped the unreachable `start < literals_start` check; the remaining `start + match_len > current_len` guard is kept.

File: zstd/src/encoding/bt/mod.rs [hunk 2]
```diff
@@ -102,4 +139,632 @@ impl BtMatcher {
...
+            let start = literals_start.saturating_add(lit_len);
+            if start < *literals_start || start + match_len > current_len {
+                continue;
+            }
+            let literals = &current[*literals_start..start];
+            let (off_base, next_reps) =
+                Self::encode_offset_with_reps(item.offset, literals.len(), *reps);
+            opt_state.update_stats(literals.len(), literals, off_base, match_len);
+            *reps = next_reps;
+            *literals_start = start + match_len;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/bt/mod.rs]: @coderabbitai Fixed — `update_plan_stats_segment` now uses `checked_add` on both `start` and `end`, drops the unreachable `start < *literals_start` branch, and reuses the computed `end` for `*literals_start = end`.

File: zstd/src/encoding/hc/mod.rs [hunk 3]
```diff
@@ -47,4 +60,484 @@ impl HcMatcher {
...
+                }
+                break;
+            }
+            cur = next;
+            if candidate_abs < table.history_abs_start || candidate_abs >= abs_pos {
+                continue;
+            }
+            buf[filled] = candidate_abs;
+            filled += 1;
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/hc/mod.rs:162]: @coderabbitai Fixed — decoded via `MatchTable::stored_abs_position_fast(cur, position_base, index_shift)` so rebased entries are correct, and capped both loop bound and fill bound at `MAX_HC_SEARCH_DEPTH` (BT-mode `search_depth` can exceed it).

File: zstd/src/encoding/match_table/storage.rs [hunk 4]
```diff
@@ -133,4 +498,1001 @@ impl MatchTable {
...
+        self.begin_rebase();
+        let history_start = self.history_abs_start;
+        // Rebuild only the already-inserted prefix. The caller inserts abs_pos
+        // immediately after this, and later positions are added in-order.
+        if self.uses_bt {
+            self.replay_history_for_rebase_bt(history_start, abs_pos);
+        } else {
+            self.replay_history_for_rebase_hc(history_start, abs_pos);
+        }
+        self.next_to_update3 = self.next_to_update3.max(abs_pos);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_table/storage.rs]: @coderabbitai Fixed — `rebase_positions_cold` now resets `next_to_update3 = history_start` and calls `update_hash3_until(abs_pos)` when `hash3_log != 0`, so the HC3 side table is replayed alongside the main hash/chain. Added regression test `rebase_positions_cold_rebuilds_hash3_for_btultra2`.

File: zstd/src/encoding/match_table/storage.rs [hunk 5]
```diff
@@ -133,4 +498,1049 @@ impl MatchTable {
...
+            }
+            let literals = &current[literals_start..start];
+            handle_sequence(Sequence::Triple {
+                literals,
+                offset: item.offset as usize,
+                match_len,
+            });
+            encode_offset_with_history(item.offset, literals.len() as u32, &mut self.offset_hist);
+            literals_start = start + match_len;
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_table/storage.rs:1396]: @coderabbitai Fixed — `emit_optimal_plan` now uses `checked_add` for both `start` and `end`, with `literals_start = end` reusing the computed value. Matches the `update_plan_stats_segment` pattern exactly.

File: zstd/src/encoding/match_table/storage.rs [hunk 6]
```diff
@@ -98,7 +126,344 @@ impl MatchTable {
...
+        if self.hash_table.is_empty() {
+            self.hash_table = alloc::vec![HC_EMPTY; 1 << self.hash_log];
+            let hash3_size = if self.hash3_log == 0 {
+                0
+            } else {
+                1 << self.hash3_log
+            };
+            self.hash3_table = alloc::vec![HC_EMPTY; hash3_size];
+            self.chain_table = alloc::vec![HC_EMPTY; 1 << self.chain_log];
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_table/storage.rs:256]: @coderabbitai Fixed — `ensure_tables` now checks each of the three tables' `len()` against the size implied by its `*_log` and reallocates per-table on mismatch. Removes the silent-UB landmine for any caller that touches `hash_log` / `chain_log` / `hash3_log` without going through `configure()` first.
